### PR TITLE
Users prmittal dtauiwithazure

### PR DESCRIPTION
**Pull request by prmittal**

### CODE REVIEW

File: Tasks/DeployVisualStudioTestAgent/DeployTestAgent.ps1 [hunk 1]
```diff
@@ -65,14 +72,10 @@ if ( [string]::IsNullOrEmpty($personalAccessToken))
...
-}
-else
-{
-    Invoke-DeployTestAgent -MachineNames $testMachines -UserName $machineUserName -Password $machinePassword -PowerShellPort 5985 -TestMachineGroup $testMachineGroup -RunAsProcess $runAsProcess -LogonAutomatically $logonAutomatically -DisableScreenSaver $disableScreenSaver -AgentLocation $agentLocation -UpdateTestAgent $updateTestAgent -InstallAgentScriptLocation $installAgentScriptLocation -ConfigureTestAgentScriptLocation $configureTestAgentScriptLocation -CheckAgentInstallationScriptLocation $checkAgentInstallationScriptLocation -downloadTestAgentScriptLocation $downloadTestAgentScriptLocation -Connection $connection -PersonalAccessToken $personalAccessToken -DataCollectionOnly $isDataCollectionOnly
-}
+Write-Verbose "Calling Register Environment cmdlet"
```
Comment (acesiddhu) [inline Tasks/DeployVisualStudioTestAgent/DeployTestAgent.ps1]: when agent is old this will certainly fail here as it wontfind new args. Check Kavipriya PR on how it shuld be done.. Also on prem people can upgrade their agents. what is the story for hosted agents? will dta task wont work till hosted agents are updated?

File: Tasks/RunDistributedTests/RunDistributedTests.ps1 [hunk 2]
```diff
@@ -49,7 +49,7 @@ Write-Verbose "Calling Invoke-RunDistributedTests"
 $runTitleMemberExists = CmdletHasMember "RunTitle"
 if($runTitleMemberExists)
 {
-    Invoke-RunDistributedTests -TestMachineGroup $testMachineGroup -SourceFilter $sourcefilters -TestCaseFilter $testFilterCriteria -RunSettingsPath $runSettingsFile -Platform $platform -Configuration $configuration -CodeCoverageEnabled $codeCoverageEnabled -TestRunParams $overrideRunParams -TestDropLocation $dropLocation -Connection $connection -TestConfiguration $testConfigurations -AutMachineGroup $autMachineGroup -UnregisterTestAgentScriptLocation $unregisterTestAgentScriptLocation -TestRunTitle $testRunTitle
+    Invoke-RunDistributedTests -TestMachineGroup $testMachineGroup -SourceFilter $sourcefilters -TestCaseFilter $testFilterCriteria -RunSettingsPath $runSettingsFile -Platform $platform -Configuration $configuration -CodeCoverageEnabled $codeCoverageEnabled -TestRunParams $overrideRunParams -TestDropLocation $dropLocation -Connection $connection -TestConfiguration $testConfigurations -AutMachineGroup $autMachineGroup -UnregisterTestAgentScriptLocation $unregisterTestAgentScriptLocation -TestRunTitle $testRunTitle -TaskContext $distributedTaskContext
```
Comment (acesiddhu) [inline Tasks/RunDistributedTests/RunDistributedTests.ps1]: same comment here

**Comment by acesiddhu** [comment]
waiting

File: Tasks/DeployVisualStudioTestAgent/DeployTestAgent.ps1 [hunk 3]
```diff
@@ -14,10 +18,20 @@ param(
...
+    return $vssEndpoint.Authorization.Parameters.AccessToken
+}
+
 Write-Verbose "Entering script DeployTestAgent.ps1"
-Write-Verbose "testMachineGroup = $testMachineGroup"
+Write-Verbose "testMachineInput = $testMachineGroup"
+Write-Verbose "WinRmProtocal = $winRmProtocol"
 Write-Verbose "resourceFilteringMethod = $resourceFilteringMethod" -Verbose
-Write-Verbose "testMachines = $testMachines"
+Write-Verbose "filter testMachines = $testMachines"
```
Comment (allendm-msft) [inline Tasks/DeployVisualStudioTestAgent/DeployTestAgent.ps1]: why filter?

File: Tasks/DeployVisualStudioTestAgent/DeployTestAgent.ps1 [hunk 4]
```diff
@@ -57,22 +74,34 @@ $connection = Get-VssConnection -TaskContext $distributedTaskContext
 
 Write-Verbose "Getting Personal Access Token for the Run"
 $vssEndPoint = Get-ServiceEndPoint -Context $distributedTaskContext -Name "SystemVssConnection"
-$personalAccessToken = $vssEndpoint.Authorization.Parameters.AccessToken
+$personalAccessToken = Get-PersonalAccessToken $vssEndpoint
 
-if ( [string]::IsNullOrEmpty($personalAccessToken))
+if (!$personalAccessToken)
```
Comment (allendm-msft) [inline Tasks/DeployVisualStudioTestAgent/DeployTestAgent.ps1]: why this change?

File: Tasks/DeployVisualStudioTestAgent/DeployTestAgent.ps1 [hunk 5]
```diff
@@ -57,22 +74,34 @@ $connection = Get-VssConnection -TaskContext $distributedTaskContext
...
+    Write-Verbose "Leaving script DeployTestAgent.ps1"
 }
 else
 {
-    Invoke-DeployTestAgent -MachineNames $testMachines -UserName $machineUserName -Password $machinePassword -PowerShellPort 5985 -TestMachineGroup $testMachineGroup -RunAsProcess $runAsProcess -LogonAutomatically $logonAutomatically -DisableScreenSaver $disableScreenSaver -AgentLocation $agentLocation -UpdateTestAgent $updateTestAgent -InstallAgentScriptLocation $installAgentScriptLocation -ConfigureTestAgentScriptLocation $configureTestAgentScriptLocation -CheckAgentInstallationScriptLocation $checkAgentInstallationScriptLocation -downloadTestAgentScriptLocation $downloadTestAgentScriptLocation -Connection $connection -PersonalAccessToken $personalAccessToken -DataCollectionOnly $isDataCollectionOnly
+    Write-Verbose "Calling old Invoke-DeployTestAgent"
```
Comment (allendm-msft) [inline Tasks/DeployVisualStudioTestAgent/DeployTestAgent.ps1]: what is a user comprehend from "old"?

File: Tasks/DeployVisualStudioTestAgent/VerifyTestMachinesAreInUse.ps1 [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+	$testAgentPath = "HKLM:\SOFTWARE\Microsoft\VisualStudio\{0}\EnterpriseTools\QualityTools\Agent" -f $Version
+	
+	if (-not (Test-Path $testAgentPath))
+	{
+		$testAgentPath = "HKLM:\SOFTWARE\Wow6432Node\Microsoft\VisualStudio\{0}\EnterpriseTools\QualityTools\Agent" -f $Version
+	}
+	
+	if (-not (Test-Path $testAgentPath))
+	{
+        Write-Verbose "Test Agent is not running as Path doesn't exist" -verbose
```
Comment (allendm-msft) [inline Tasks/DeployVisualStudioTestAgent/VerifyTestMachinesAreInUse.ps1]: indent

Comment (allendm-msft) [inline Tasks/DeployVisualStudioTestAgent/VerifyTestMachinesAreInUse.ps1]: user comments need to have a more understandable error and the corrective action.

Comment (allendm-msft) [inline Tasks/DeployVisualStudioTestAgent/VerifyTestMachinesAreInUse.ps1]: isnt this error instead of verbose?

File: Tasks/DeployVisualStudioTestAgent/VerifyTestMachinesAreInUse.ps1 [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+		return $false
+	}
+	
+	$testAgentTfsUrl = (Get-ItemProperty $testAgentPath -ErrorAction SilentlyContinue).TfsUrl
+	if (($testAgentTfsUrl -eq $null) -or ($testAgentTfsUrl.Length -eq 0))
```
Comment (allendm-msft) [inline Tasks/DeployVisualStudioTestAgent/VerifyTestMachinesAreInUse.ps1]: isnullorwhitespace

File: Tasks/DeployVisualStudioTestAgent/VerifyTestMachinesAreInUse.ps1 [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+    {
+        Write-Verbose "Test Agent is not running as TfsUrl is empty" -verbose
```
Comment (allendm-msft) [inline Tasks/DeployVisualStudioTestAgent/VerifyTestMachinesAreInUse.ps1]: same comments as above

File: Tasks/DeployVisualStudioTestAgent/VerifyTestMachinesAreInUse.ps1 [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+    $targetKeys = @()      # New array
+    foreach ($key in $keys)
+    {
+      $targetKeys += [decimal] $key
+    }
+
+    return $targetKeys
+}
+
+Check-TestAgentIsRunning -Version "14.0"
```
Comment (allendm-msft) [inline Tasks/DeployVisualStudioTestAgent/VerifyTestMachinesAreInUse.ps1]: indent the file properly

File: Tasks/DeployVisualStudioTestAgent/VerifyTestMachinesAreInUse.ps1 [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+	{
+		$regPath = "HKLM:\SOFTWARE\Wow6432Node\Microsoft\DevDiv\vstf\Servicing"
+	}
+	if (-not (Test-Path $regPath))
+	{
+		return $null
+	}
+	
+	$keys = Get-Item $regPath | %{$_.GetSubKeyNames()} -ErrorAction SilentlyContinue
+	$Version = Get-SubKeysInFloatFormat $keys | Sort-Object -Descending | Select-Object -First 1
```
Comment (allendm-msft) [inline Tasks/DeployVisualStudioTestAgent/VerifyTestMachinesAreInUse.ps1]: small v

File: Tasks/DeployVisualStudioTestAgent/VerifyTestMachinesAreInUse.ps1 [hunk 11]
```diff
@@ -0,0 +1,67 @@
+function Check-TestAgentIsRunning([string] $Version)
+{
+	$AvlVersion = Locate-TestVersion
```
Comment (allendm-msft) [inline Tasks/DeployVisualStudioTestAgent/VerifyTestMachinesAreInUse.ps1]: you are returning $null from this function and not handling it.

File: Tasks/DeployVisualStudioTestAgent/task.json [hunk 12]
```diff
@@ -13,12 +13,12 @@
...
         "Major": 1,
         "Minor": 0,
-        "Patch": 20
+        "Patch": 21
     },
     "demands": [
 
     ],
-    "minimumAgentVersion": "1.88.0",
+    "minimumAgentVersion": "1.92.0",
```
Comment (allendm-msft) [inline Tasks/DeployVisualStudioTestAgent/task.json]: should be 95 now right?

File: Tasks/DeployVisualStudioTestAgent/task.json [hunk 13]
```diff
@@ -37,17 +37,58 @@
...
+		{
             "name": "testMachineGroup",
-            "type": "pickList",
-            "label": "Test Machine Group / Azure Resource Group",
+            "type": "multiLine",
+            "label": "Machines",
             "defaultValue": "",
             "required": true,
-            "helpMarkDown": "Name of the Test Machine Group or Azure Resource Group on which tests need to be run.",
+            "helpMarkDown": "Provide a comma separated list of machine IP addresses or FQDNs along with ports. Port is defaulted based on the selected protocol. <br>Eg: dbserver.fabrikam.com,dbserver_int.fabrikam.com:5986,192.168.12.34:5986 <br>Or provide output variable of other tasks. Eg: $(variableName) <br>Or provide a Machine Group Name.",
```
Comment (allendm-msft) [inline Tasks/DeployVisualStudioTestAgent/task.json]: place examples in `` quotes to get a verdana/console font

File: Tasks/DeployVisualStudioTestAgent/task.json [hunk 14]
```diff
@@ -37,17 +37,58 @@
...
+            "name": "winRmProtocol",
+            "type": "radio",
+            "label": "Protocol",
+            "required": false,
+            "defaultValue": "Https",
+            "options": {
+                "Http": "HTTP",
+                "Https": "HTTPS"
+            },
+            "helpMarkDown": "Select the protocol to use for the WinRM connection with the machine(s). Default is HTTPS.",
```
Comment (allendm-msft) [inline Tasks/DeployVisualStudioTestAgent/task.json]: ``

File: Tasks/RunDistributedTests/RunDistributedTests.ps1 [hunk 15]
```diff
@@ -70,45 +69,19 @@ if([int]::TryParse($testPlan, [ref]$testPlanId)){}
...
-   }
-   if([string]::Equals($testSelection, "testPlan"))
-   {
-     if($testSelectionMemberExists)
-     {
-        Invoke-RunDistributedTests -TestMachineGroup $testMachineGroup -SourceFilter $sourcefilters -TestCaseFilter $testFilterCriteria -RunSettingsPath $runSettingsFile -Platform $platform -Configuration $configuration -CodeCoverageEnabled $codeCoverageEnabled -TestRunParams $overrideRunParams -TestDropLocation $dropLocation -Connection $connection -TestConfiguration $testConfigurations -AutMachineGroup $autMachineGroup -UnregisterTestAgentScriptLocation $unregisterTestAgentScriptLocation -TestSelection $testSelection -TestPlan $testPlanId -TestSuites $testSuites -TestConfig $testConfigurationId
-     }
-     else
-     {
+    if([string]::Equals($testSelection, "testPlan"))
```
Comment (allendm-msft) [inline Tasks/RunDistributedTests/RunDistributedTests.ps1]: ignore case

**Comment by allendm-msft** [comment]
waiting

File: Tasks/DeployVisualStudioTestAgent/DeployTestAgent.ps1 (hunk 3 above)
Comment (nigurr) [inline Tasks/DeployVisualStudioTestAgent/DeployTestAgent.ps1]: The initial parameter name is went wrong. testMachineGroup = input of test machines
testMachines = filter value. To make it easy, I added that

File: Tasks/DeployVisualStudioTestAgent/DeployTestAgent.ps1 (hunk 4 above)
Comment (nigurr) [inline Tasks/DeployVisualStudioTestAgent/DeployTestAgent.ps1]: To write PS test easily and able to mack

File: Tasks/DeployVisualStudioTestAgent/DeployTestAgent.ps1 (hunk 5 above)
Comment (nigurr) [inline Tasks/DeployVisualStudioTestAgent/DeployTestAgent.ps1]: It's verbose right. I thought it's not meant for user.

File: Tasks/DeployVisualStudioTestAgent/VerifyTestMachinesAreInUse.ps1 (hunk 6 above)
Comment (nigurr) [inline Tasks/DeployVisualStudioTestAgent/VerifyTestMachinesAreInUse.ps1]: it's not error. it's pre-check if any machine is already running any tests.

File: Tasks/DeployVisualStudioTestAgent/VerifyTestMachinesAreInUse.ps1 (hunk 11 above)
Comment (nigurr) [inline Tasks/DeployVisualStudioTestAgent/VerifyTestMachinesAreInUse.ps1]: if($AvlVersion) -> null or empty check in ps

**Comment by nigurr** [comment]
added CR changes

**Comment by acesiddhu** [comment]
looks fine
